galley/pkg/source/kube/dynamic/source.go

```diff
@@ -77,80 +73,65 @@ func New(
-	s.stateLock.Lock()
-	defer s.stateLock.Unlock()
-
-	if s.stopCh != nil {
-		return fmt.Errorf("already synchronizing resources: name='%s', gv='%v'",
-			s.spec.Singular, s.spec.GroupVersion())
-	}
-	if handler == nil {
-		return fmt.Errorf("invalid event handler")
-	}
-
-	log.Scope.Debugf("Starting source for %s(%v)", s.spec.Singular, s.spec.GroupVersion())
+	return s.worker.Start(func(ctx context.Context) {
+		if handler == nil {
+			panic("dynamic source provided with nil event handler")
+		}
 
-	s.stopCh = make(chan struct{})
-	s.handler = handler
+		log.Scope.Debugf("Starting source for %s(%v)", s.spec.Singular, s.spec.GroupVersion())
 
-	s.informer = cache.NewSharedIndexInformer(
-		&cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (k8sRuntime.Object, error) {
-				return s.resourceClient.List(options)
+		s.handler = handler
+		s.informer = cache.NewSharedIndexInformer(
+			&cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (k8sRuntime.Object, error) {
+					return s.resourceClient.List(options)
+				},
+				WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+					options.Watch = true
+					return s.resourceClient.Watch(options)
+				},
 			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				options.Watch = true
-				return s.resourceClient.Watch(options)
+			&unstructured.Unstructured{},
+			s.resyncPeriod,
+			cache.Indexers{})
+
+		s.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) { s.handleEvent(resource.Added, obj) },
+			UpdateFunc: func(old, new interface{}) {
+				newRes := new.(*unstructured.Unstructured)
+				oldRes := old.(*unstructured.Unstructured)
+				if newRes.GetResourceVersion() == oldRes.GetResourceVersion() {
+					// Periodic resync will send update events for all known resources.
+					// Two different versions of the same resource will always have different RVs.
+					return
+				}
+				s.handleEvent(resource.Updated, new)
 			},
-		},
-		&unstructured.Unstructured{},
-		s.resyncPeriod,
-		cache.Indexers{})
-
-	s.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) { s.handleEvent(resource.Added, obj) },
-		UpdateFunc: func(old, new interface{}) {
-			newRes := new.(*unstructured.Unstructured)
-			oldRes := old.(*unstructured.Unstructured)
-			if newRes.GetResourceVersion() == oldRes.GetResourceVersion() {
-				// Periodic resync will send update events for all known resources.
-				// Two different versions of the same resource will always have different RVs.
-				return
-			}
-			s.handleEvent(resource.Updated, new)
-		},
-		DeleteFunc: func(obj interface{}) { s.handleEvent(resource.Deleted, obj) },
-	})
+			DeleteFunc: func(obj interface{}) { s.handleEvent(resource.Deleted, obj) },
+		})
 
-	// Start CRD shared informer background process.
-	go s.informer.Run(s.stopCh)
+		// Start CRD shared informer background process.
+		go s.informer.Run(ctx.Done())
```

Comment:
I think you can do this as the last thing, within the same go-routine, and have the same effect, right? That should save one go-routine per listener.

Reply:
good idea ... done.